Feat: 본문 판별 알고리즘 개선(tistory)

dlgnswk opened the pull request:
## 개요
### Resolves: #12
<!---- 변경 사항 및 관련 이슈가 있다면 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
`tistory`는 유저의 커스텀 스킨으로 인해 시멘틱 태그를 통한 구분이 불가능한 경우가 있어 별도의 판별 알고리즘을 작성하였습니다.
`tistory`는 조사해본 결과 크게 두가지 구조로 구분할 수 있었습니다.
1. 시멘틱을 지키고 있는 구조
`main`, `article`, `section` 등

2. 시멘틱을 지키지 않고 특정 단어를 사용해 클래스명으로 식별하는 구조
`class="article-xxx"`, `class="content_xxx"`, `class="area_xxx"` 등

최대한 많은 커스텀 스킨을 판별할 수 있도록 확장 가능성을 염두하여 작성하였습니다.

## 코드 변경 사항
- 중첩 `try..catch`문을 통해 기존에 작성된 시맨틱 아티클 판별 알고리즘을 통해 먼저 판별가능한지 확인하고 유효하지 않아 `error`를 던지는 경우 `tistory` 판별 알고리즘으로 판별할 수 있도록 하였습니다.
https://github.com/team-sticky-252/readim-server/blob/35d2b7adcfac409b26878cda1fdc8e7171535e77/src/app.controller.js#L123-L129
- 시멘틱하지 않은 경우를 판별할 수 있는 `className`의 공통된 단어를 상수화하여 확장 가능하도록 작성하였습니다.
https://github.com/team-sticky-252/readim-server/blob/35d2b7adcfac409b26878cda1fdc8e7171535e77/src/app.service.js#L12-L21
- 본문 내에서 사용자가 읽는 범위로 판단되지 않는 `className`의 공통된 단어를 상수화하여 확장 가능하도록 작성하였습니다.
https://github.com/team-sticky-252/readim-server/blob/35d2b7adcfac409b26878cda1fdc8e7171535e77/src/app.service.js#L8-L9
- 시멘틱한 스킨과 시멘틱을 준수하지 않은 스킨에 대한 경우에 대해 테스트 코드를 작성하였습니다.
- 기존 `convertCodeTagToText()` 함수는 본문의 태그를 순회하며 `pre` 태그내에 `code` 블럭의 단어를 구분하여 문자열로 반환하도록 작성되어 있었습니다. 이를 통해 카멜케이스로 작성된 `className`을 단어별로 구분하여 총 단어수를 집계하였는데 인라인 코드는 카멜케이스에 대응하고 있지 않아 일관성 있는 단어수 집계를 위해 대상 범위를 `pre` 태그에서 `code`로 변경하였습니다.


## 기타 전달 사항
조사한 범위내에서 본문으로 판별할 수 있는 `css selector`는 다음과 같습니다.
- `id` : `article-view`
- `className` : `tt_article_useless_p_margin`, `contents_style`, `area_view`, `contents_style`

본문 내에서 판별대상으로 제외되어야 하는 `css selector`는 다음과 같습니다.
- `className`: `container_postbtn`, `#post_button_group`


## PR Checklist
- [x] 주어진 Task의 요구사항을 모두 작성했습니다.
- [x] 작성한 PR을 다시 한번 더 검토하였습니다.
- [x] merge 할 브랜치의 위치를 확인하였습니다.

dlgnswk committed:
Feat: tistory 대응 본문 판별 로직 추가

- tistory.com 에 대응하여 본문을 판별할 수 있는 로직을 작성하였습니다.
- class name을 통해 필요, 불필요 요소를 구분하는 로직을 작성하였습니다.

dlgnswk committed:
Refactor: convertCodeTagToText 함수의 적용대상 변경

- 단어수 체크의 일관성을 위해 인라인 코드 영역을 포함하여 단어를 분리하였습니다.

Pangho04 commented on src/app.controller.js:
```diff
@@ -101,4 +101,50 @@ export class AppController {
     });
   }
+
+  @Get("/tistory")
```
API 요청 URL에 대해서 합의가 필요할 것 같습니다

dlgnswk replied:
팀원들과 합의한 결과 하나의 호출로직으로 공통 알고리즘으로 본문 판별을 먼저 수행하고, 판별이 불가능한 경우 이후에 url 별로 분기하여 해당 url에 맞는 알고리즘을 수행할 수 있도록 하였습니다.

Pangho04 reviewed (changes requested):
수정 사항 확인 부탁드립니다!

JieunZoeyOh commented on src/app.service.js:
```diff
@@ -5,7 +5,20 @@ import { HttpError, HttpStatus } from "./utils/http";
+  /(button|btn|nav|footer|summary|author|comment|sns|share|card|activity|relate|another|reply|sr-only|post-switch|loof|revenue|act|tag|direction|og|notice)/i;
+const EXCLUDED_ID_REGEX = /(paging|player)/i;
 const MAIN_CONTENT_TAGS_REGEX = /^(main|article|section)$/i;
+const MAIN_CONTENT_CSS_REGEX =
+  /^(tt_article_useless_p_margin|area_view|contents_style|contents_style|article_skin)$/i;
```
```suggestion
const MAIN_CONTENT_CSS_REGEX =
  /(tt_article_useless_p_margin|area_view|contents_style|contents_style|article_skin)/i;
```
정규표현식의 ^는 문자열의 시작을 나타내고, $는 문자열의 끝을 의미합니다.
즉, 현재 작성된 정규표현식은 전체 문자열이 괄호 안의 항목 중 하나와 정확히 일치해야 함을 나타내고 있습니다.

물론 class에 괄호 안의 항목 중 하나만 사용한 경우에는 문제가 없겠지만 다른 클래스명이 함께 사용된다면 의도와 다르게 본문으로 판별하지 못할 것 같습니다.
ex) `<div class="tt_article_useless_p_margin foo"> ... </div>`

JieunZoeyOh commented on src/app.service.js:
```diff
@@ -67,10 +80,20 @@ export class AppService {
     while (stack.length > 0) {
       const currentElement = stack.pop();
       const tagName = currentElement.tagName.toLowerCase();
+      const { className } = currentElement;
+      const { id } = currentElement;
```
```suggestion
      const { className, id } = currentElement;
```
한 줄로 줄일 수 있을 것 같습니다!

JieunZoeyOh reviewed (changes requested):
재훈님 티스토리의 경우 변수가 많아 어려웠을 텐데 짧은 시간 내에 최대한 자료조사 많이 하시려고 하고 또 그거에 맞게 대응하시느라 고생 많으셨습니다!
리뷰 코멘트 확인 부탁드립니다!

JieunZoeyOh commented on src/app.controller.js:
```diff
@@ -101,4 +101,50 @@ export class AppController {
+
+  @Get("/tistory")
+  @Bind(Req(), Res())
+  async getTistoryArticle(request, response) {
```
전체적으로 중복되는 로직을 service에 메서드로 분리하면 중복을 제거할 수 있을 것 같아요!

```js
async getMainContent(url) {
  const { bodyElement } = await this.getHtmlElement(url);

  try {
    return this.getSemanticMainContent(bodyElement);
  } catch (error) {
    return this.getTistoryMainContent(bodyElement);
  }
};
```

dlgnswk committed:
Feat: 본문에서 제외되어야 할 요소의 class 명 추가

dlgnswk committed:
Refactor: 클래스명 매칭의 정확도를 위한 정규표현식 수정

dlgnswk committed:
Refactor: 가독성을 고려한 코드 수정

dlgnswk committed:
Refactor: API 호출 로직의 중복 제거 및 통합을 위한 구조 재조정

- URL 별 API 호출 로직의 중복을 제거하고 단일 호출로 통합하였습니다.
- app.service.js를 통해 각 URL에 따른 분기 처리를 추가하였습니다.
- 메서드 변경에 따라 테스트 내용을 수정하였습니다.

JieunZoeyOh commented on src/app.service.js:
```diff
@@ -338,4 +332,22 @@ export class AppService {
+      if (url.includes("tistory.com")) {
+        return this.getTistoryMainContent(bodyElement);
+      }
+    }
```
velog와 tistory가 아닌 경우 getSemanticMainContent에서 발생한 에러를 getMainContent를 호출한 곳으로 다시 throw 하는 것이 좋을 것 같습니다!

Pangho04 reviewed (approved):
고생하셨습니다!

dlgnswk committed:
Refactor: 함수 작성 순서 변경 및 에러 처리 추가

dlgnswk committed:
Fix: 읽기 속도 분석 로직에서 필요한 URL 인자 누락 수정
